google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher2.java

```diff
@@ -0,0 +1,278 @@
+                  builder.clear();
+                  more = drainAckNacks(builder);
+                  // If both are 0, there's nothing to send.
+                  if (builder.getAckIdsCount() != 0
+                      || builder.getModifyDeadlineAckIdsCount() != 0) {
+                    sendRequest(builder.build());
+                  }
+                }
+              }
+            },
+            100,
+            100,
+            TimeUnit.MILLISECONDS);
+
+    this.extensionJob =
+        systemExecutor.scheduleAtFixedRate(
+            new Runnable() {
+              @Override
+              public void run() {
+                List<String> extendIds = new ArrayList<>(extensionSet.keySet());
+                for (List<String> batch : Lists.partition(extendIds, MAX_CHANGE_PER_REQUEST)) {
+                  sendRequest(
+                      StreamingPullRequest.newBuilder()
+                          .addAllModifyDeadlineAckIds(batch)
+                          .addAllModifyDeadlineSeconds(
+                              Collections.nCopies(batch.size(), DEADLINE_EXTENSION_SEC))
+                          .build());
+                }
+              }
+            },
+            KEEPALIVE_SEC,
+            KEEPALIVE_SEC,
+            TimeUnit.SECONDS);
+  }
+
+  synchronized void shutdown() {
+    ackNackJob.cancel(false);
+    extensionJob.cancel(false);
+  }
+
+  private boolean drainAckNacks(StreamingPullRequest.Builder builder) {
+    while (builder.getAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      String id = idsToAck.poll();
+      if (id == null) {
+        break;
+      }
+      builder.addAckIds(id);
+    }
+
+    while (builder.getModifyDeadlineAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      String id = idsToNack.poll();
+      if (id == null) {
+        break;
+      }
+      builder.addModifyDeadlineAckIds(id);
+    }
+    builder.addAllModifyDeadlineSeconds(
+        Collections.nCopies(builder.getModifyDeadlineAckIdsCount(), 0));
+
+    return builder.getAckIdsCount() == MAX_CHANGE_PER_REQUEST
+        || builder.getModifyDeadlineAckIdsCount() == MAX_CHANGE_PER_REQUEST;
+  }
+
+  void processReceivedMessages(List<ReceivedMessage> messages, Runnable doneCallback) {
+    if (messages.isEmpty()) {
+      doneCallback.run();
+      return;
+    }
+
+    StreamingPullRequest.Builder builder =
+        StreamingPullRequest.newBuilder()
+            .addAllModifyDeadlineSeconds(
+                Collections.nCopies(messages.size(), DEADLINE_EXTENSION_SEC));
+    for (ReceivedMessage message : messages) {
+      builder.addModifyDeadlineAckIds(message.getAckId());
+    }
+    sendRequest(builder.build());
+
+    AtomicInteger remaining = new AtomicInteger(messages.size());
+    for (ReceivedMessage message : messages) {
+      workQueue.add(new WorkItem(message, remaining, doneCallback));
+      extensionSet.put(message.getAckId(), true);
+    }
+    workMessages();
+  }
+
+  private synchronized void workMessages() {
```

Comment:
This being synchronized is a little worrisome, since this is a hot path. Since the load test shows no problem, I think we should leave this as-is. I can think of a somewhat-lock-free way to implement this, but it's rather complicated. We can do that if this turns out to be a bottleneck.